### ISSUE TITLE
Add X close button next to each open script file name.

### ISSUE THREAD

**Issue by wyattbiker** [godotengine/godot-proposals#218]
**Describe the project you are working on:**
Any
**Describe the problem or limitation you are having in your project:**
Closing open scripts is not as convenient as with modern editors. 

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Add X close button next to each open script file name. Would make it easier, faster and less confusing than using CMD-W (for Mac) or (Right Click and choosing close).

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![image](https://user-images.githubusercontent.com/930478/68537506-3b389200-0333-11ea-94f7-587c741775f1.png)

**Describe implementation detail for your proposal (in code), if possible:**
unsure

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Will be used very often if not exclusively as opposed to CMD-W (on Mac) or RightClick and choose  close

**Is there a reason why this should be core and not an add-on in the asset library?:**
should be part of editor.

**Comment by Calinou** [comment]
You can middle-click scripts in the list to close them. This also works with scene tabs.

**Comment by golddotasksquestions** [comment]
I find the current workflow also really confusing and would really welcome the x. It's also really annoying that scripts of deleted nodes still linger around in this list and you have to manually close them.
Even much worse: They even linger around if you delete the .gd file from the file browser!! No indication that any chance has happened (3.1). In 3.2 there is an indication because the name disappears from the list, but not the script itself, not the place in the list, and neither the little cog symbol. You then have to rightclick and select "close" on the empty thing in the list. It's really weird and confusing.

**Comment by KoBeWi** [comment]
@golddotasksquestions What you describe is a bug ([which is already reported btw](https://github.com/godotengine/godot/issues/33313)).

**Comment by IronicallySerious** [comment]
Giving close buttons to every script open in the list of open scripts is quite common though.

Most text editors have it and even larger IDEs have this. However, there is a small detail that those programs have a horizontally stacked list of open files (instead of a horizontal one that we find in Godot), which makes it very obvious that closing the file is really not going to affect the file contents (as opposed to the indication that the file will be deleted/altered).

**Comment by IronicallySerious** [comment]
>Since this is already available, I'd prefer not adding a close button which could be misclicked.

@Calinou Even though the middle-click shortcut is present everywhere, a close button is just as prevalent too. I'd say keeping things similar to what people see normally would be a good idea.

**Comment by MCrafterzz** [comment]
Middle clicking feels strange so I would prefer a small button

**Comment by wyattbiker** [comment]
Mac users trackpad or Mighty Mouse don't have middle buttons. Also the argument you can click on X by accident, not sure if it holds water :).  The close button will notify you to save your code if it was modified. Worse thing will happen is to use Reopen Closed Script. It's right there on the File menu.

**Comment by Jummit** [comment]
I would like this, but only when the x is only shown for scripts that are currently being edited or hovered over:
![image](https://user-images.githubusercontent.com/28286961/68681241-60abe280-0563-11ea-9996-21c19bc3901a.png)

**Comment by Calinou** [comment]
@Jummit This sounds fine to me. Maybe we could turn the "X" into a dot when the file has unsaved changes (to replace the `(*)` marker next to the filename)?

**Comment by Jummit** [comment]
Isn't that the case with scene tabs?

**Comment by Calinou** [comment]
@Jummit Not yet, though I've been meaning to do it:

![Current modified scene tab](https://user-images.githubusercontent.com/180032/68681490-ca2bf100-0563-11ea-9b7d-d59d85783550.png)

![Proposed modified scene tab](https://user-images.githubusercontent.com/180032/68681820-6229da80-0564-11ea-8608-800f05f97033.png)

**Comment by golddotasksquestions** [comment]
The problem with the middle mouse button is that there is no indicator this functionality exists (not even in the context menu). An x is a clear indicator.

**Comment by wyattbiker** [comment]
@Calinou,

What @Jummit is suggesting (only show X for open files) makes sense in Visual Code. Visual Code (and Sublime), allows you to single click on any file in the **Project** and see a `Preview` of the file. Click away to another file and the Preview closes and re-opens the next one. Preview would be nice feature as well in Godot, but that it is another proposal and far more complex :)

The X's shown in the VCode image, is because he double clicked (or modified the Preview) which opens the script in a new tab. For Godot  to keep things simple and convenient an X button or if you want move the modified (*) to flush right and when hovered turn it into X should be sufficient.

My suggestions for future enhancements is install Sublime (or VSCode) and check out how they do it. X button is close, **o** is modified but turns to X on hover. Italicized name is a Preview tab. Very elegant.

![image](https://user-images.githubusercontent.com/930478/68722919-03c12400-0585-11ea-84bd-5c85e034efd6.png)

**Comment by Paulb23** [comment]
Related engine issue: godotengine/godot/issues/3774

Similar to VSCode, QTCreator also only shows the close button when hovered. I personally favour this workflow.

With that said, I don't see any reason not to have an option similar to scene tabs, that allows the user to have the close button displayed at all times.

**Comment by Xrayez** [comment]
> Also the argument you can click on X by accident, not sure if it holds water 🙂. The close button will notify you to save your code if it was modified. Worse thing will happen is to use Reopen Closed Script. It's right there on the File menu.

Yes, also any unsaved scripts/text files (which were likely deleted or moved) are now covered by godotengine/godot#40647, so not a big deal in my eyes.

I always tend to click on the text part as well, so not sure if I'd click on the "close" button by accident. If I do, I'd just <kbd>Ctrl + Shift + T</kbd> and it's back.
